index.bs

```diff
@@ -4675,6 +4675,45 @@ Instead, in step three, the comparison on the host is relaxed to accept hosts on
 : Authenticator extension output
 :: None.
 
+## FIDO AppID Exclusion Extension (appidExclude) ## {#sctn-appid-exclude-extension}
+
+This registration extension allows [=[WRPS]=] to exclude authenticators which contain specified credentials that were created with FIDO APIs.
+
+During a transition from legacy FIDO JavaScript APIs, a [=[RP]=] may have a population of users with legacy credentials already registered. The [appid](#sctn-appid-extension) extension allows the sign-in flow to be transitioned smoothly but, when transitioning the registration flow, the [excludeCredentials](#dom-publickeycredentialcreationoptions-excludecredentials) field will not be effective in excluding authenticators with legacy credentials because its contents are taken to be WebAuthn credentials. This extension directs [=client platforms=] to consider the contents of [excludeCredentials](#dom-publickeycredentialcreationoptions-excludecredentials) as both WebAuthn and legacy FIDO credentials.
+
+: Extension identifier
+:: `appidExclude`
+
+: Operation applicability
+:: [=registration extension|Registration=]
+
+: Client extension input
+:: A single USVString specifying a FIDO |AppID|.
+    <xmp class="idl">
+    partial dictionary AuthenticationExtensionsClientInputs {
+      USVString appidExclude;
+    };
+    </xmp>
+
+: Client extension processing
+::  1. Let |facetId| be the result of passing the caller's [=origin=] to the FIDO algorithm for [=determining the FacetID of a calling application=].
+    1. Let |appId| be the extension input.
+    1. Pass |facetId| and |appId| to the FIDO algorithm for [=determining if a caller's FacetID is authorized for an AppID=]. If that algorithm rejects |appId| then return a "{{SecurityError}}" {{DOMException}}.
+    1. For each authenticator that supports the U2F protocol, if |excludeCredentialDescriptorList| is not empty then, before invoking [=authenticatorMakeCredential=], attempt to obtain a U2F assertion (without user-presence), for each element of |excludeCredentialDescriptorList|, using the SHA-256 hash of |appId| as the application parameter. If unsuccessful, continue with invoking [=authenticatorMakeCredential=]. Otherwise, cease normal processing of this authenticator and indicate in a platform-specific manner that the authenticator is inapplicable. For example, this could be in the form of UI, or could involve requesting user-consent from the authenticator and, upon receipt, treating it as if the authenticator had returned {{InvalidStateError}}.
```

Comment:
I had to dig around to figure out how this works. IIUC, it seems that steps 1 - 3 (above) probably ought to occur around step 19 of [\[\[create\]\]()](https://w3c.github.io/webauthn/#sctn-createCredential), and step 4 (above) ought to occur either after or during Step 20.5 of [\[\[create\]\]()](https://w3c.github.io/webauthn/#sctn-createCredential).  Perhaps we oughta add links into those places in \[\[create\]\]()](https://w3c.github.io/webauthn/#sctn-createCredential) from the above. I have some alg-smithing suggestions for step 4 yet to write up. I can help with crafting this PR if you wish.

Reply:
Indeed, there were a lot of commas in there. I've restructured this a bit, see if you like it better. You also have direct push access to this repo. (I.e. `git remote add agl git@github.com:agl/webauthn.git` then `git push agl HEAD:appidexclude` to push the current branch to this PR.)